src/Components/Typeahead/Base.purs

```diff
@@ -15,97 +16,125 @@ import Effect.Aff.Class (class MonadAff)
+type Slot f item id = H.Slot (Query f item) (Output f item) id
 
-type DSL pq f item m =
-  H.ParentDSL
-    (StateStore pq f item m)
-    (Query pq f item m)
-    (ChildQuery pq f item m)
-    ChildSlot
-    (Message pq f item)
-    m
+type Component f item m = H.Component HH.HTML (Query f item) (Input f item m) (Output f item) m
+type ComponentHTML f item m = H.ComponentHTML (Action f item m) (ChildSlots f item) m
+type ComponentRender f item m = State f item m -> ComponentHTML f item m
+type ComponentM f item m a = H.HalogenM (StateStore f item m) (Action f item m) (ChildSlots f item) (Output f item) m a
 
-type StateStore pq f item m = Store
-  (State f item m)
-  (H.ParentHTML (Query pq f item m) (ChildQuery pq f item m) ChildSlot m)
-
-type State f item m =
-  { items :: RemoteData String (Array item)
-  , selected :: f item
-  , search :: String
+type StateRow f item m =
+  ( items :: RemoteData String (Array item) -- NOTE pst.items, Parent(Typeahead)
   , insertable :: Insertable item
   , keepOpen :: Boolean
   , itemToObject :: item -> Object String
-  , ops :: Operations f item
-  , debounceTime :: Maybe Milliseconds
   , async :: Maybe (String -> m (RemoteData String (Array item)))
-  }
 
-type Input pq f item m =
+  , ops :: Operations f item
+  , config :: { debounceTime :: Maybe Milliseconds }
+  , selected :: f item
+  , fuzzyItems :: Array (Fuzzy item) -- NOTE cst.items, Child(Select)
+  )
+
+type State f item m = Record (StateRow f item m)
+
+type StateStore f item m = Store (State f item m) (ComponentHTML f item m)
+
+type Input f item m =
   { items :: RemoteData String (Array item)
   , insertable :: Insertable item
   , keepOpen :: Boolean
   , itemToObject :: item -> Object String
-  , debounceTime :: Maybe Milliseconds
   , async :: Maybe (String -> m (RemoteData String (Array item)))
-  , render :: -- All rendering will happen in Select, but with access to parent state
-      State f item m
-      -> Select.State (Fuzzy item)
-      -> Select.ComponentHTML (Query pq f item m) (Fuzzy item)
+
+  , debounceTime :: Maybe Milliseconds
+  , render :: CompositeComponentRender f item m
   }
 
-data Query pq f item m a
-  = Remove item a
-  | RemoveAll a
-  | TriggerFocus a
-  | Synchronize a
-  | Search String a
-  | HandleSelect (Select.Message (Query pq f item m) (Fuzzy item)) a
-  | GetSelected (f item -> a)
+data Action f item (m :: Type -> Type)
+  = PassingOutput (Output f item)
+  | ReceiveRender (Input f item m)
+
+data EmbeddedAction (f :: Type -> Type) item (m :: Type -> Type)
+  = Initialize
+  | Remove item
+  | RemoveAll
+  -- | Receive CompositeInput
+-- NOTE internal actions, moved to Util functions
+  -- | Synchronize a
+-- NOTE deprecated
+  -- | Search String
+  -- | TriggerFocus a
+  -- | HandleSelect (Select.Message (Query pq f item m) (Fuzzy item)) a
+  -- | AndThen (Query pq f item m Unit) (Query pq f item m Unit) a
+  -- | Raise (pq Unit) a
```

Comment:
Making a note here. If it's temporary and removed by the end of the review, ignore it.

Can we kill this dead code?